src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -246,96 +253,111 @@ object RuleUtils {
+      // If there are unhandled appended files, we need to create additional plans
+      // by the following steps:
+      // Step 1) Generate a plan(planForAppended) from the current plan to read
+      //   the appended files similar to indexPlan.
+      // Step 2) If Shuffle is required, perform shuffle for the plan.
+      // Step 3) Merge both indexPlan and planForAppended by using [[BucketUnion]] or [[Union]].
+      // For more details, see https://github.com/microsoft/hyperspace/issues/150.
+
+      val planForAppended =
+        transformPlanWithAppendedFiles(spark, index.schema, plan, filesAppended)
+      if (useBucketSpec) {
+        // If Bucketing information of the index is used to read the index data, we need to
+        // shuffle the appended data in the same way to correctly merge with bucketed index data.
+
+        // Clear sortColumnNames as BucketUnion does not keep the sort order within a bucket.
+        val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
+
+        // Merge index plan & newly shuffled plan by using bucket-aware union.
+        BucketUnion(
+          Seq(indexPlan, shuffleUsingIndexSpec(bucketSpec, planForAppended)),
+          bucketSpec)
+      } else {
+        // If bucketing is not necessary (e.g. FilterIndexRule), we use [[Union]] to merge
+        // the appended data without additional shuffle.
+        Union(indexPlan, planForAppended)
+      }
     } else {
-      replacedPlan
+      indexPlan
     }
   }
 
   /**
-   * Get complement plan for an index with appended data.
-   *
-   * This method consists of the following steps
-   * 1) Get a plan from originalPlan by replacing data location with appended data
-   * 2) On-the-fly shuffle for the appended data, using indexedColumns & numBuckets.
-   *   - Shuffle is located before Project to utilize Push-down Filters
-   *     - Shuffle => Project => Filter => Relation
-   *   - if Project filters indexedColumns, then Shuffle should be located after the node
-   *     - Project => Shuffle => Filter => Relation
-   * 3) Bucket-aware union both indexPlan and complementPlan to avoid repartitioning index data
+   * Transform the current plan to read the given appended source files.
    *
-   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   * The result will be merged with the plan which is reading index data
+   * by using [[BucketUnion]] or [[Union]].
    *
    * @param spark Spark session.
-   * @param index Index used in replacement plan.
+   * @param indexSchema Index schema used for the output.
    * @param originalPlan Original plan.
-   * @param planWithIndex Replaced plan with index.
    * @param filesAppended Appended files to the source relation.
-   * @return Consolidated plan of planWithIndex and a complement plan for filesAppended
+   * @return Linear logical plan for appended files.
    */
-  private def getComplementIndexPlan(
+  private def transformPlanWithAppendedFiles(
```

Comment:
```suggestion
  private def transformPlanToReadFromAppendedFiles(
```

What do you think? Might also be consistent with our use of `To` in the method names above.